server/src/main/java/org/candlepin/model/EntitlementCurator.java

```diff
@@ -987,4 +924,132 @@ public Entitlement findUpstreamEntitlementForStack(Consumer consumer, String sta
 
         return (Entitlement) activeNowQuery.uniqueResult();
     }
+
+    /**
+     * Defines the queries used to find the collection of modifier entitlements of
+     * a given set of entitlements.
+     */
+    private class DependentEntitlementQuery {
+
+        private final String PPP1_TABLE_PARAM = "PPP1_TABLE";
+
+        private final String BASE_QUERY = "SELECT DISTINCT e2.id " +
+            // Required entitlement
+            "FROM cp_entitlement e1 " +
+            // Required entitlement => required pool
+            "JOIN " + PPP1_TABLE_PARAM + " ppp1 ON ppp1.pool_id = e1.pool_id " +
+            // Required pool => required product
+            "JOIN cp2_products p ON p.uuid = ppp1.product_uuid " +
+            // Required product => conditional content
+            "JOIN cp2_content_modified_products cmp ON cmp.element = p.product_id " +
+            // Conditional content => dependent product
+            "JOIN cp2_product_content pc ON pc.content_uuid = cmp.content_uuid " +
+            // Dependent product => dependent pool
+            "JOIN cp2_pool_provided_products ppp2 ON ppp2.product_uuid = pc.product_uuid " +
+            // Dependent pool => dependent entitlement
+            "JOIN cp_entitlement e2 ON e2.pool_id = ppp2.pool_id " +
+            "WHERE e1.consumer_id = e2.consumer_id " +
+            "  AND e1.id != e2.id " +
+            "  AND e2.dirty = false " +
+            "  AND e1.id IN (:entitlement_ids)" +
+            "  AND e2.id NOT IN (:entitlement_ids)";
+
+        private EntityManager em;
+        private int blockSize;
+
+        public DependentEntitlementQuery(EntityManager em, int blockSize) {
+            this.em = em;
+            this.blockSize = blockSize;
```

Comment:
Block size should not be a parameter here -- this object should determine the proper block size based on the query it's going to be executing.